frontend/server/src/main/java/org/pytorch/serve/grpcimpl/InferenceImpl.java

```diff
@@ -79,42 +82,122 @@ public void predictions(
             Status status,
             Exception e,
-            String description) {
-        responseObserver.onError(
-                status.withDescription(e.getMessage())
-                        .augmentDescription(
-                                description == null ? e.getClass().getCanonicalName() : description)
-                        .withCause(e)
-                        .asRuntimeException());
+            String description,
+            WorkerCommands workerCmd) {
+        if (workerCmd == WorkerCommands.STREAMPREDICT2) {
+            com.google.rpc.Status rpcStatus =
+                    com.google.rpc.Status.newBuilder()
+                            .setCode(status.getCode().value())
+                            .setMessage(e.getMessage())
+                            .addDetails(
+                                    Any.pack(
+                                            ErrorInfo.newBuilder()
+                                                    .setReason(
+                                                            description == null
+                                                                    ? e.getClass()
+                                                                            .getCanonicalName()
+                                                                    : description)
+                                                    .build()))
+                            .build();
+            PredictionResponse response =
+                    PredictionResponse.newBuilder().setStatus(rpcStatus).build();
+            responseObserver.onNext(response);
+        } else {
+            responseObserver.onError(
+                    status.withDescription(e.getMessage())
+                            .augmentDescription(
+                                    description == null
+                                            ? e.getClass().getCanonicalName()
+                                            : description)
+                            .withCause(e)
+                            .asRuntimeException());
+        }
     }
 
     private void prediction(
             PredictionsRequest request,
             StreamObserver<PredictionResponse> responseObserver,
             WorkerCommands workerCmd) {
-        ((ServerCallStreamObserver<PredictionResponse>) responseObserver)
-                .setOnCancelHandler(
-                        () -> {
-                            logger.warn("grpc client call already cancelled");
-                            responseObserver.onError(
-                                    io.grpc.Status.CANCELLED
-                                            .withDescription("call already cancelled")
-                                            .asRuntimeException());
-                        });
+        if (workerCmd != WorkerCommands.STREAMPREDICT2) {
+            ((ServerCallStreamObserver<PredictionResponse>) responseObserver)
+                    .setOnCancelHandler(
```

Comment:
streamPredictions2 is bi-direction streaming; streamPredictions is uni-direction (ie. streaming response). 
`setOnCancelHandler` needs to be set at the beginning of function streamPredictions2 and streamPredictions. Function prediction can be directly called by function streamPredictions; However, it can only be called streaming request `onNext` function.